Reusing a texture in an image map

I'm trying to use one texture for multiple images. My texture consists of all of a bunch of icons packed together in order to save on the number of textures. However, with the glutin/glium backend, I can't seem to have an image map of references to `glium::texture::Texture2d`s like so:
```
let image_map = image_map! {
    (ids.image1, &icons),
    (ids.image2, &icons)
};
```
The error I get when I use references is this:
```
error[E0271]: type mismatch resolving `<&conrod::glium::Texture2d as std::ops::Deref>::Target == conrod::glium::texture::TextureAny`
   --> src/main.rs:150:16
    |
150 |       renderer.fill(&display, primitives, &image_map);
    |                ^^^^ expected struct `conrod::glium::Texture2d`, found struct `conrod::glium::texture::TextureAny`
    |
    = note: expected type `conrod::glium::Texture2d`
    = note:    found type `conrod::glium::texture::TextureAny`
    = note: required because of the requirements on the impl of `conrod::backend::glium::TextureDimensions` for `&conrod::glium::Texture2d`

error[E0277]: the trait bound `for<'a> conrod::glium::uniforms::Sampler<'a, &conrod::glium::Texture2d>: conrod::glium::uniforms::AsUniformValue` is not satisfied
   --> src/main.rs:154:16
    |
154 |       renderer.draw(&display, &mut target, &image_map).unwrap();
    |                ^^^^
    |
    = help: the following implementations were found:
    = help:   <conrod::glium::uniforms::Sampler<'a, conrod::glium::texture::Texture1d> as conrod::glium::uniforms::AsUniformValue>
    = help:   <conrod::glium::uniforms::Sampler<'a, conrod::glium::texture::CompressedTexture1d> as conrod::glium::uniforms::AsUniformValue>
    = help:   <conrod::glium::uniforms::Sampler<'a, conrod::glium::texture::SrgbTexture1d> as conrod::glium::uniforms::AsUniformValue>
    = help:   <conrod::glium::uniforms::Sampler<'a, conrod::glium::texture::CompressedSrgbTexture1d> as conrod::glium::uniforms::AsUniformValue>
    = help: and 55 others

```
Is there any way I can use one texture for multiple images?

--- a/examples/all_piston_window.rs
+++ b/examples/all_piston_window.rs
@@ -34,9 +34,6 @@ mod feature {
                 .build()
                 .unwrap();
 
-        // A demonstration of some state that we'd like to control with the App.
-        let mut app = support::DemoApp::new();
-
         // construct our `Ui`.
         let mut ui = conrod::UiBuilder::new([WIDTH as f64, HEIGHT as f64])
             .theme(support::theme())
@@ -75,7 +72,11 @@ mod feature {
 
         // Create our `conrod::image::Map` which describes each of our widget->image mappings.
         // In our case we only have one image, however the macro may be used to list multiple.
-        let image_map = support::image_map(&ids, rust_logo);
+        let mut image_map = conrod::image::Map::new();
+        let rust_logo = image_map.insert(rust_logo);
+
+        // A demonstration of some state that we'd like to control with the App.
+        let mut app = support::DemoApp::new(rust_logo);
 
         // Poll events from the window.
         while let Some(event) = window.next() {

--- a/examples/all_winit_gfx.rs
+++ b/examples/all_winit_gfx.rs
@@ -208,9 +208,6 @@ mod feature {
         // Compile GL program
         let pso = factory.create_pipeline_simple(VERTEX_SHADER, FRAGMENT_SHADER, pipe::new()).unwrap();
 
-        // Demonstration app state that we'll control with our conrod GUI.
-        let mut app = support::DemoApp::new();
-
         // Create Ui and Ids of widgets to instantiate
         let mut ui = conrod::UiBuilder::new([WIN_W as f64, WIN_H as f64]).theme(support::theme()).build();
         let ids = support::Ids::new(ui.widget_id_generator());
@@ -240,6 +237,15 @@ mod feature {
             (cache, texture, texture_view)
         };
 
+        // FIXME: We don't yet load the rust logo, so just insert nothing for now so we can get an
+        // identifier used to construct the DemoApp. This should be changed to *actually* load a
+        // gfx texture for the rust logo and insert it into the map.
+        let mut image_map = conrod::image::Map::new();
+        let rust_logo = image_map.insert(());
+
+        // Demonstration app state that we'll control with our conrod GUI.
+        let mut app = support::DemoApp::new(rust_logo);
+
         // Event loop
         let mut events = window.poll_events();
 
@@ -266,7 +272,7 @@ mod feature {
                         },
                         render::PrimitiveKind::Lines { color, cap, thickness, points } => {
                         },
-                        render::PrimitiveKind::Image { color, source_rect } => {
+                        render::PrimitiveKind::Image { image_id, color, source_rect } => {
                         },
                         render::PrimitiveKind::Text { color, text, font_id } => {
                             let positioned_glyphs = text.positioned_glyphs(dpi_factor);

--- a/examples/all_winit_glium.rs
+++ b/examples/all_winit_glium.rs
@@ -34,9 +34,6 @@ mod feature {
             .build_glium()
             .unwrap();
 
-        // A demonstration of some app state that we want to control with the conrod GUI.
-        let mut app = support::DemoApp::new();
-
         // Construct our `Ui`.
         let mut ui = conrod::UiBuilder::new([WIN_W as f64, WIN_H as f64]).theme(support::theme()).build();
 
@@ -59,7 +56,11 @@ mod feature {
             texture
         }
 
-        let image_map = support::image_map(&ids, load_rust_logo(&display));
+        let mut image_map = conrod::image::Map::new();
+        let rust_logo = image_map.insert(load_rust_logo(&display));
+
+        // A demonstration of some app state that we want to control with the conrod GUI.
+        let mut app = support::DemoApp::new(rust_logo);
 
         // A type used for converting `conrod::render::Primitives` into `Command`s that can be used
         // for drawing to the glium `Surface`.

--- a/examples/image.rs
+++ b/examples/image.rs
@@ -46,9 +46,8 @@ mod feature {
         // In our case we only have one image, however the macro may be used to list multiple.
         let rust_logo = load_rust_logo(&display);
         let (w, h) = (rust_logo.get_width(), rust_logo.get_height().unwrap());
-        let image_map = image_map! {
-            (ids.rust_logo, rust_logo),
-        };
+        let mut image_map = conrod::image::Map::new();
+        let rust_logo = image_map.insert(rust_logo);
 
         // Poll events from the window.
         let mut event_loop = support::EventLoop::new();
@@ -77,7 +76,7 @@ mod feature {
                 // Draw a light blue background.
                 widget::Canvas::new().color(color::LIGHT_BLUE).set(ids.background, ui);
                 // Instantiate the `Image` at its full size in the middle of the window.
-                widget::Image::new().w_h(w as f64, h as f64).middle().set(ids.rust_logo, ui);
+                widget::Image::new(rust_logo).w_h(w as f64, h as f64).middle().set(ids.rust_logo, ui);
             }
 
             // Render the `Ui` and then display it on the screen.

--- a/examples/image_button.rs
+++ b/examples/image_button.rs
@@ -57,9 +57,8 @@ mod feature {
         // In our case we only have one image, however the macro may be used to list multiple.
         let rust_logo = load_rust_logo(&display);
         let (w, h) = (rust_logo.get_width(), rust_logo.get_height().unwrap());
-        let image_map = image_map! {
-            (ids.rust_logo, rust_logo),
-        };
+        let mut image_map = conrod::image::Map::new();
+        let rust_logo = image_map.insert(rust_logo);
 
         // We'll change the background colour with the image button.
         let mut bg_color = conrod::color::LIGHT_BLUE;
@@ -96,7 +95,7 @@ mod feature {
                     .set(ids.canvas, ui);
 
                 // Button widget example button.
-                if widget::Button::image(ids.rust_logo)
+                if widget::Button::image(rust_logo)
                     .w_h(w as conrod::Scalar, h as conrod::Scalar)
                     .middle_of(ids.canvas)
                     .color(color::TRANSPARENT)

--- a/examples/support/mod.rs
+++ b/examples/support/mod.rs
@@ -28,17 +28,19 @@ pub struct DemoApp {
     ball_xy: conrod::Point,
     ball_color: conrod::Color,
     sine_frequency: f32,
+    rust_logo: conrod::image::Id,
 }
 
 
 impl DemoApp {
 
     /// Simple constructor for the `DemoApp`.
-    pub fn new() -> Self {
+    pub fn new(rust_logo: conrod::image::Id) -> Self {
         DemoApp {
             ball_xy: [0.0, 0.0],
             ball_color: conrod::color::WHITE,
             sine_frequency: 1.0,
+            rust_logo: rust_logo,
         }
     }
 
@@ -64,14 +66,6 @@ pub fn theme() -> conrod::Theme {
         widget_styling: std::collections::HashMap::new(),
         mouse_drag_threshold: 0.0,
         double_click_threshold: std::time::Duration::from_millis(500),
-    }
-}
-
-
-/// Create an image map that maps the `ids.rust_logo` to the `rust_logo` image.
-pub fn image_map<T>(ids: &Ids, rust_logo: T) -> conrod::image::Map<T> {
-    image_map! {
-        (ids.rust_logo, rust_logo)
     }
 }
 
@@ -244,7 +238,7 @@ pub fn gui(ui: &mut conrod::UiCell, ids: &Ids, app: &mut DemoApp) {
         .set(ids.image_title, ui);
 
     const LOGO_SIDE: conrod::Scalar = 144.0;
-    widget::Image::new()
+    widget::Image::new(app.rust_logo)
         .w_h(LOGO_SIDE, LOGO_SIDE)
         .down(60.0)
         .align_middle_x_of(ids.canvas)

--- a/src/backend/glium.rs
+++ b/src/backend/glium.rs
@@ -8,7 +8,6 @@ use image;
 use render;
 use std;
 use text;
-use widget;
 
 
 /// A `Command` describing a step in the drawing process.
@@ -27,13 +26,13 @@ pub enum Command<'a> {
 pub enum Draw<'a> {
     /// A range of vertices representing triangles textured with the image in the
     /// image_map at the given `widget::Id`.
-    Image(widget::Id, &'a [Vertex]),
+    Image(image::Id, &'a [Vertex]),
     /// A range of vertices representing plain triangles.
     Plain(&'a [Vertex]),
 }
 
 enum PreparedCommand {
-    Image(widget::Id, std::ops::Range<usize>),
+    Image(image::Id, std::ops::Range<usize>),
     Plain(std::ops::Range<usize>),
     Scizzor(glium::Rect),
 }
@@ -339,7 +338,7 @@ impl Renderer {
         vertices.clear();
 
         enum State {
-            Image { id: widget::Id, start: usize },
+            Image { image_id: image::Id, start: usize },
             Plain { start: usize },
         }
 
@@ -351,8 +350,8 @@ impl Renderer {
             () => {
                 match current_state {
                     State::Plain { .. } => (),
-                    State::Image { id, start } => {
-                        commands.push(PreparedCommand::Image(id, start..vertices.len()));
+                    State::Image { image_id, start } => {
+                        commands.push(PreparedCommand::Image(image_id, start..vertices.len()));
                         current_state = State::Plain { start: vertices.len() };
                     },
                 }
@@ -393,7 +392,7 @@ impl Renderer {
 
         // Draw each primitive in order of depth.
         while let Some(primitive) = primitives.next_primitive() {
-            let render::Primitive { id, kind, scizzor, rect } = primitive;
+            let render::Primitive { kind, scizzor, rect, .. } = primitive;
 
             // Check for a `Scizzor` command.
             let new_scizzor = rect_to_glium_rect(scizzor);
@@ -402,8 +401,8 @@ impl Renderer {
                 match current_state {
                     State::Plain { start } =>
                         commands.push(PreparedCommand::Plain(start..vertices.len())),
-                    State::Image { id, start } =>
-                        commands.push(PreparedCommand::Image(id, start..vertices.len())),
+                    State::Image { image_id, start } =>
+                        commands.push(PreparedCommand::Image(image_id, start..vertices.len())),
                 }
 
                 // Update the scizzor and produce a command.
@@ -603,31 +602,37 @@ impl Renderer {
                     }
                 },
 
-                render::PrimitiveKind::Image { color, source_rect } => {
+                render::PrimitiveKind::Image { image_id, color, source_rect } => {
 
                     // Switch to the `Image` state for this image if we're not in it already.
-                    let widget_id = id;
+                    let new_image_id = image_id;
                     match current_state {
 
                         // If we're already in the drawing mode for this image, we're done.
-                        State::Image { id, .. } if id == widget_id => (),
+                        State::Image { image_id, .. } if image_id == new_image_id => (),
 
                         // If we were in the `Plain` drawing state, switch to Image drawing state.
                         State::Plain { start } => {
                             commands.push(PreparedCommand::Plain(start..vertices.len()));
-                            current_state = State::Image { id: id, start: vertices.len() };
+                            current_state = State::Image {
+                                image_id: new_image_id,
+                                start: vertices.len(),
+                            };
                         },
 
                         // If we were drawing a different image, switch state to draw *this* image.
-                        State::Image { id, start } => {
-                            commands.push(PreparedCommand::Image(id, start..vertices.len()));
-                            current_state = State::Image { id: widget_id, start: vertices.len() };
+                        State::Image { image_id, start } => {
+                            commands.push(PreparedCommand::Image(image_id, start..vertices.len()));
+                            current_state = State::Image {
+                                image_id: new_image_id,
+                                start: vertices.len(),
+                            };
                         },
                     }
 
                     let color = gamma_srgb_to_linear(color.unwrap_or(color::WHITE).to_fsa());
 
-                    let (image_w, image_h) = image_map.get(&id).unwrap().dimensions();
+                    let (image_w, image_h) = image_map.get(&image_id).unwrap().dimensions();
                     let (image_w, image_h) = (image_w as Scalar, image_h as Scalar);
 
                     // Get the sides of the source rectangle as uv coordinates.
@@ -683,8 +688,8 @@ impl Renderer {
         match current_state {
             State::Plain { start } =>
                 commands.push(PreparedCommand::Plain(start..vertices.len())),
-            State::Image { id, start } =>
-                commands.push(PreparedCommand::Image(id, start..vertices.len())),
+            State::Image { image_id, start } =>
+                commands.push(PreparedCommand::Image(image_id, start..vertices.len())),
         }
     }
 
@@ -725,9 +730,9 @@ impl Renderer {
 
                     // Draw an image whose texture data lies within the `image_map` at the
                     // given `id`.
-                    Draw::Image(id, slice) => {
+                    Draw::Image(image_id, slice) => {
                         let vertex_buffer = glium::VertexBuffer::new(facade, slice).unwrap();
-                        let image = image_map.get(&id).unwrap();
+                        let image = image_map.get(&image_id).unwrap();
                         let image_uniforms = uniform! {
                             tex: glium::uniforms::Sampler::new(image)
                                 .wrap_function(glium::uniforms::SamplerWrapFunction::Clamp)

--- a/src/backend/piston/draw.rs
+++ b/src/backend/piston/draw.rs
@@ -88,7 +88,7 @@ pub fn primitive<'a, Img, G, T, C, F>(
           C: FnMut(&mut G, &mut T, text::rt::Rect<u32>, &[u8]),
           F: FnMut(&Img) -> &T,
 {
-    let render::Primitive { id, kind, scizzor, rect } = primitive;
+    let render::Primitive { kind, scizzor, rect, .. } = primitive;
     let view_size = context.get_view_size();
     // Translate the `context` to suit conrod's orientation (middle (0, 0), y pointing upwards).
     let context = context.trans(view_size[0] / 2.0, view_size[1] / 2.0).scale(1.0, -1.0);
@@ -187,8 +187,8 @@ pub fn primitive<'a, Img, G, T, C, F>(
                                               graphics);
         },
 
-        render::PrimitiveKind::Image { color, source_rect } => {
-            if let Some(img) = image_map.get(&id) {
+        render::PrimitiveKind::Image { image_id, color, source_rect } => {
+            if let Some(img) = image_map.get(&image_id) {
                 let mut image = piston_graphics::image::Image::new();
                 image.color = color.map(|c| c.to_fsa());
                 if let Some(source_rect) = source_rect {

--- a/src/image.rs
+++ b/src/image.rs
@@ -3,7 +3,14 @@
 //! - [Map](./struct.Map.html)
 
 use std;
-use widget;
+
+/// Unique image identifier.
+///
+/// Throughout conrod, images are referred to via their unique `Id`. By referring to images via
+/// `Id`s, conrod can remain agnostic of the actual image or texture types used to represent each
+/// image.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Id(u32);
 
 /// A type used to map the `widget::Id` of `Image` widgets to their associated `Img` data.
 ///
@@ -18,6 +25,7 @@ use widget;
 /// };
 /// ```
 pub struct Map<Img> {
+    next_index: u32,
     map: HashMap<Img>,
     /// Whether or not the `image::Map` will trigger a redraw the next time `Ui::draw` is called.
     ///
@@ -26,7 +34,13 @@ pub struct Map<Img> {
 }
 
 /// The type of `std::collections::HashMap` used within the `image::Map`.
-pub type HashMap<Img> = std::collections::HashMap<widget::Id, Img>;
+pub type HashMap<Img> = std::collections::HashMap<Id, Img>;
+
+/// An iterator yielding an `Id` for each new `Img` inserted into the `Map` via the `extend`
+/// method.
+pub struct NewIds {
+    index_range: std::ops::Range<u32>,
+}
 
 
 impl<Img> std::ops::Deref for Map<Img> {
@@ -42,19 +56,18 @@ impl<Img> Map<Img> {
     /// Construct a new, empty `image::Map`.
     pub fn new() -> Self {
         Map {
+            next_index: 0,
             map: std::collections::HashMap::new(),
             trigger_redraw: std::cell::Cell::new(true),
         }
     }
 
-
     // Calling any of the following methods will trigger a redraw when using `Ui::draw_if_changed`.
-
 
     /// Uniquely borrow the `Img` associated with the given widget.
     ///
     /// Note: Calling this will trigger a redraw the next time `Ui::draw_if_changed` is called.
-    pub fn get_mut(&mut self, id: widget::Id) -> Option<&mut Img> {
+    pub fn get_mut(&mut self, id: Id) -> Option<&mut Img> {
         self.trigger_redraw.set(true);
         self.map.get_mut(&id)
     }
@@ -67,33 +80,46 @@ impl<Img> Map<Img> {
     /// from the map and returned.
     ///
     /// Note: Calling this will trigger a redraw the next time `Ui::draw_if_changed` is called.
-    pub fn insert(&mut self, id: widget::Id, img: Img) -> Option<Img> {
+    pub fn insert(&mut self, img: Img) -> Id {
         self.trigger_redraw.set(true);
-        self.map.insert(id, img)
+        let index = self.next_index;
+        self.next_index = index.wrapping_add(1);
+        let id = Id(index);
+        self.map.insert(id, img);
+        id
     }
 
-}
-
-impl<Img> std::iter::Extend<(widget::Id, Img)> for Map<Img> {
-    fn extend<I>(&mut self, mappings: I)
-        where I: IntoIterator<Item=(widget::Id, Img)>,
+    /// Insert each of the images yielded by the given iterator and produce an iterator yielding
+    /// their generated `Ids` in the same order.
+    ///
+    /// Note: Calling this will trigger a redraw the next time `Ui::draw_if_changed` is called.
+    pub fn extend<I>(&mut self, images: I) -> NewIds
+        where I: IntoIterator<Item=Img>,
     {
         self.trigger_redraw.set(true);
-        self.map.extend(mappings.into_iter().map(|(id, img)| (id, img)));
+        let start_index = self.next_index;
+        let mut end_index = start_index;
+        for image in images {
+            self.map.insert(Id(end_index), image);
+            end_index += 1;
+        }
+        NewIds { index_range: start_index..end_index }
+    }
+
+}
+
+impl Iterator for NewIds {
+    type Item = Id;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.index_range.next().map(|i| Id(i))
+    }
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.len()))
     }
 }
 
-
-/// A macro for simplifying the instantiation of an `image::Map`.
-///
-/// See the [**Map**](./image/struct.Map.html) documentation for an example.
-#[macro_export]
-macro_rules! image_map {
-    ($(($idx:expr, $img:expr)),* $(,)*) => {{
-        let mut map = $crate::image::Map::new();
-        $(
-            map.insert($idx, $img);
-        )*
-        map
-    }};
+impl ExactSizeIterator for NewIds {
+    fn len(&self) -> usize {
+        self.index_range.len()
+    }
 }

--- a/src/render.rs
+++ b/src/render.rs
@@ -11,6 +11,7 @@
 
 use {Align, Color, Dimensions, FontSize, Point, Rect, Scalar};
 use graph::{self, Graph};
+use image;
 use std;
 use text;
 use theme::Theme;
@@ -131,6 +132,8 @@ pub enum PrimitiveKind<'a> {
 
     /// A single `Image`, produced by the primitive `Image` widget.
     Image {
+        /// The unique identifier of the image that will be drawn.
+        image_id: image::Id,
         /// When `Some`, colours the `Image`. When `None`, the `Image` uses its regular colours.
         color: Option<Color>,
         /// The area of the texture that will be drawn to the `Image`'s `Rect`.
@@ -205,6 +208,7 @@ enum OwnedPrimitiveKind {
         point_range: std::ops::Range<usize>,
     },
     Image {
+        image_id: image::Id,
         color: Option<Color>,
         source_rect: Option<Rect>,
     },
@@ -516,6 +520,7 @@ impl<'a> Primitives<'a> {
                     let color = style.maybe_color(theme);
                     let kind = PrimitiveKind::Image {
                         color: color,
+                        image_id: state.image_id,
                         source_rect: state.src_rect,
                     };
                     return Some(new_primitive(id, kind, scizzor, rect));
@@ -580,8 +585,9 @@ impl<'a> Primitives<'a> {
                     primitives.push(new(kind));
                 },
 
-                PrimitiveKind::Image { color, source_rect } => {
+                PrimitiveKind::Image { image_id, color, source_rect } => {
                     let kind = OwnedPrimitiveKind::Image {
+                        image_id: image_id,
                         color: color,
                         source_rect: source_rect,
                     };
@@ -758,8 +764,9 @@ impl<'a> WalkOwnedPrimitives<'a> {
                     new(kind)
                 },
 
-                OwnedPrimitiveKind::Image { color, source_rect } => {
+                OwnedPrimitiveKind::Image { image_id, color, source_rect } => {
                     let kind = PrimitiveKind::Image {
+                        image_id: image_id,
                         color: color,
                         source_rect: source_rect,
                     };

--- a/src/widget/button.rs
+++ b/src/widget/button.rs
@@ -14,6 +14,7 @@ use {
     UiCell,
     Widget,
 };
+use image;
 use text;
 use widget;
 
@@ -52,16 +53,22 @@ widget_style!{
 }
 
 widget_ids! {
-    struct Ids {
+    /// Identifiers for a "flat" button.
+    #[allow(missing_docs, missing_copy_implementations)]
+    pub struct FlatIds {
         rectangle,
         label,
     }
 }
 
-/// The style of the `Button`, either `Flat` or `Image`.
-pub trait Show: Sized {
-    /// Display the unique styling of the `Button`.
-    fn show(self, _button_id: widget::Id, _ui: &mut UiCell) {}
+widget_ids! {
+    /// Identifiers for an image button.
+    #[allow(missing_docs, missing_copy_implementations)]
+    pub struct ImageIds {
+        rectangle,
+        label,
+        image,
+    }
 }
 
 /// The `Button` simply displays a flat color.
@@ -72,7 +79,7 @@ pub struct Flat;
 #[derive(Copy, Clone)]
 pub struct Image {
     /// The id of the `Image` to be used.
-    pub id: widget::Id,
+    pub image_id: image::Id,
     /// If `Some`, maps the image's luminance to this `Color`.
     pub color: ImageColor,
     /// The rectangular area of the original source image that should be displayed.
@@ -90,12 +97,6 @@ pub enum ImageColor {
     WithFeedback(Color),
     /// The image's regular color will be used.
     None,
-}
-
-
-/// Represents the state of the Button widget.
-pub struct State {
-    ids: Ids,
 }
 
 /// The `Event` type yielded by the `Button` widget.
@@ -128,9 +129,9 @@ impl Iterator for TimesClicked {
 impl<'a> Button<'a, Image> {
 
     /// Begin building a button displaying the given `Image` on top.
-    pub fn image(image_id: widget::Id) -> Self {
+    pub fn image(image_id: image::Id) -> Self {
         let image = Image {
-            id: image_id,
+            image_id: image_id,
             src_rect: None,
             color: ImageColor::None,
         };
@@ -217,10 +218,8 @@ impl<'a, S> Button<'a, S> {
 }
 
 
-impl<'a, S> Widget for Button<'a, S>
-    where S: Show,
-{
-    type State = State;
+impl<'a> Widget for Button<'a, Flat> {
+    type State = FlatIds;
     type Style = Style;
     type Event = TimesClicked;
 
@@ -233,9 +232,47 @@ impl<'a, S> Widget for Button<'a, S>
     }
 
     fn init_state(&self, id_gen: widget::id::Generator) -> Self::State {
-        State {
-            ids: Ids::new(id_gen),
+        FlatIds::new(id_gen)
+    }
+
+    fn style(&self) -> Style {
+        self.style.clone()
+    }
+
+    /// Update the state of the Button.
+    fn update(self, args: widget::UpdateArgs<Self>) -> Self::Event {
+        let widget::UpdateArgs { id, state, style, rect, ui, .. } = args;
+        let Button { maybe_label, .. } = self;
+
+        let (color, times_clicked) = color_and_times_clicked(id, &style, ui);
+
+        bordered_rectangle(id, state.rectangle, rect, color, style, ui);
+
+        // Label widget.
+        if let Some(l) = maybe_label {
+            label(id, state.label, state.rectangle, l, style, ui);
         }
+
+        TimesClicked(times_clicked)
+    }
+
+}
+
+impl<'a> Widget for Button<'a, Image> {
+    type State = ImageIds;
+    type Style = Style;
+    type Event = TimesClicked;
+
+    fn common(&self) -> &widget::CommonBuilder {
+        &self.common
+    }
+
+    fn common_mut(&mut self) -> &mut widget::CommonBuilder {
+        &mut self.common
+    }
+
+    fn init_state(&self, id_gen: widget::id::Generator) -> Self::State {
+        ImageIds::new(id_gen)
     }
 
     fn style(&self) -> Style {
@@ -247,61 +284,94 @@ impl<'a, S> Widget for Button<'a, S>
         let widget::UpdateArgs { id, state, style, rect, ui, .. } = args;
         let Button { show, maybe_label, .. } = self;
 
-        let (color, times_clicked) = {
-            let input = ui.widget_input(id);
-            let color = style.color(ui.theme());
-            let color = input.mouse().map_or(color, |mouse| {
-                if mouse.buttons.left().is_down() {
-                    color.clicked()
-                } else {
-                    color.highlighted()
-                }
-            });
-            let times_clicked = input.clicks().left().count() as u16;
-            (color, times_clicked)
-        };
+        let (color, times_clicked) = color_and_times_clicked(id, &style, ui);
 
-        // BorderedRectangle widget.
-        let dim = rect.dim();
-        let border = style.border(&ui.theme);
-        let border_color = style.border_color(&ui.theme);
-        widget::BorderedRectangle::new(dim)
+        bordered_rectangle(id, state.rectangle, rect, color, style, ui);
+
+        // Instantiate the image.
+        let Image { image_id, src_rect, color } = show;
+        let mut image = widget::Image::new(image_id)
             .middle_of(id)
-            .graphics_for(id)
-            .color(color)
-            .border(border)
-            .border_color(border_color)
-            .set(state.ids.rectangle, ui);
+            .wh_of(id)
+            .parent(id)
+            .graphics_for(id);
+        image.src_rect = src_rect;
+        image.style.maybe_color = match color {
+            ImageColor::Normal(color) => Some(Some(color)),
+            ImageColor::WithFeedback(color) =>
+                ui.widget_input(id).mouse()
+                    .map(|mouse| if mouse.buttons.left().is_down() {
+                        Some(color.clicked())
+                    } else {
+                        Some(color.highlighted())
+                    })
+                    .or(Some(Some(color))),
+            ImageColor::None => None,
+        };
+        image.set(state.image, ui);
 
-        // This instantiates the image widget if necessary.
-        show.show(id, ui);
-
-        // Label widget.
-        if let Some(label) = maybe_label {
-            let color = style.label_color(&ui.theme);
-            let font_size = style.label_font_size(&ui.theme);
-            let align = style.label_x_align(&ui.theme);
-            let font_id = style.label_font_id(&ui.theme).or(ui.fonts.ids().next());
-            widget::Text::new(label)
-                .and_then(font_id, widget::Text::font_id)
-                .and(|b| match align {
-                    Align::Start =>
-                        b.mid_left_with_margin_on(state.ids.rectangle, font_size as Scalar),
-                    Align::Middle =>
-                        b.middle_of(state.ids.rectangle),
-                    Align::End =>
-                        b.mid_right_with_margin_on(state.ids.rectangle, font_size as Scalar),
-                })
-                .parent(id)
-                .graphics_for(id)
-                .color(color)
-                .font_size(font_size)
-                .set(state.ids.label, ui);
+        if let Some(s) = maybe_label {
+            label(id, state.label, state.rectangle, s, style, ui);
         }
 
         TimesClicked(times_clicked)
     }
 
+}
+
+
+fn color_and_times_clicked(button_id: widget::Id, style: &Style, ui: &UiCell) -> (Color, u16) {
+    let input = ui.widget_input(button_id);
+    let color = style.color(ui.theme());
+    let color = input.mouse().map_or(color, |mouse| {
+        if mouse.buttons.left().is_down() {
+            color.clicked()
+        } else {
+            color.highlighted()
+        }
+    });
+    let times_clicked = input.clicks().left().count() as u16;
+    (color, times_clicked)
+}
+
+fn bordered_rectangle(button_id: widget::Id, rectangle_id: widget::Id,
+                      rect: Rect, color: Color, style: &Style, ui: &mut UiCell)
+{
+    // BorderedRectangle widget.
+    let dim = rect.dim();
+    let border = style.border(&ui.theme);
+    let border_color = style.border_color(&ui.theme);
+    widget::BorderedRectangle::new(dim)
+        .middle_of(button_id)
+        .graphics_for(button_id)
+        .color(color)
+        .border(border)
+        .border_color(border_color)
+        .set(rectangle_id, ui);
+}
+
+fn label(button_id: widget::Id, label_id: widget::Id, rectangle_id: widget::Id,
+         label: &str, style: &Style, ui: &mut UiCell)
+{
+    let color = style.label_color(&ui.theme);
+    let font_size = style.label_font_size(&ui.theme);
+    let align = style.label_x_align(&ui.theme);
+    let font_id = style.label_font_id(&ui.theme).or(ui.fonts.ids().next());
+    widget::Text::new(label)
+        .and_then(font_id, widget::Text::font_id)
+        .and(|b| match align {
+            Align::Start =>
+                b.mid_left_with_margin_on(rectangle_id, font_size as Scalar),
+            Align::Middle =>
+                b.middle_of(rectangle_id),
+            Align::End =>
+                b.mid_right_with_margin_on(rectangle_id, font_size as Scalar),
+        })
+        .parent(button_id)
+        .graphics_for(button_id)
+        .color(color)
+        .font_size(font_size)
+        .set(label_id, ui);
 }
 
 
@@ -321,33 +391,5 @@ impl<'a, S> Labelable<'a> for Button<'a, S> {
         label { maybe_label = Some(&'a str) }
         label_color { style.label_color = Some(Color) }
         label_font_size { style.label_font_size = Some(FontSize) }
-    }
-}
-
-
-impl Show for Flat {}
-
-impl Show for Image {
-    fn show(self, button_id: widget::Id, ui: &mut UiCell) {
-        let Image { id, src_rect, color } = self;
-        let mut image = widget::Image::new()
-            .middle_of(button_id)
-            .wh_of(button_id)
-            .parent(button_id)
-            .graphics_for(button_id);
-        image.src_rect = src_rect;
-        image.style.maybe_color = match color {
-            ImageColor::Normal(color) => Some(Some(color)),
-            ImageColor::WithFeedback(color) =>
-                ui.widget_input(button_id).mouse()
-                    .map(|mouse| if mouse.buttons.left().is_down() {
-                        Some(color.clicked())
-                    } else {
-                        Some(color.highlighted())
-                    })
-                    .or(Some(Some(color))),
-            ImageColor::None => None,
-        };
-        image.set(id, ui);
     }
 }

--- a/src/widget/primitive/image.rs
+++ b/src/widget/primitive/image.rs
@@ -1,6 +1,7 @@
 //! A simple, non-interactive widget for drawing an `Image`.
 
 use {Color, Dimension, Rect, Widget, Ui};
+use image;
 use widget;
 
 
@@ -9,6 +10,8 @@ use widget;
 pub struct Image {
     /// Data necessary and common for all widget builder types.
     pub common: widget::CommonBuilder,
+    /// The unique identifier for the image that will be drawn.
+    pub image_id: image::Id,
     /// The rectangle area of the original source image that should be used.
     pub src_rect: Option<Rect>,
     /// Unique styling.
@@ -22,6 +25,8 @@ pub struct State {
     ///
     /// If `None`, the entire image will be used.
     pub src_rect: Option<Rect>,
+    /// The unique identifier for the image's associated data that will be drawn.
+    pub image_id: image::Id,
 }
 
 widget_style!{
@@ -38,7 +43,7 @@ impl Image {
     /// Construct a new `Image`.
     ///
     /// Note that the `Image` widget does not require borrowing or owning any image data directly.
-    /// Instead, image data is stored within a `conrod::image::Map` where widget indices are mapped
+    /// Instead, image data is stored within a `conrod::image::Map` where `image::Id`s are mapped
     /// to their associated data.
     ///
     /// This is done for a few reasons:
@@ -59,9 +64,10 @@ impl Image {
     /// 1. use an enum with a variant for each type
     /// 2. use a trait object, where the trait is implemented for each of your image types or
     /// 3. use an index type which may be mapped to your various image types.
-    pub fn new() -> Self {
+    pub fn new(image_id: image::Id) -> Self {
         Image {
             common: widget::CommonBuilder::new(),
+            image_id: image_id,
             src_rect: None,
             style: Style::new(),
         }
@@ -98,6 +104,7 @@ impl Widget for Image {
     fn init_state(&self, _: widget::id::Generator) -> Self::State {
         State {
             src_rect: None,
+            image_id: self.image_id,
         }
     }
 
@@ -121,8 +128,11 @@ impl Widget for Image {
 
     fn update(self, args: widget::UpdateArgs<Self>) -> Self::Event {
         let widget::UpdateArgs { state, .. } = args;
-        let Image { src_rect, .. } = self;
+        let Image { image_id, src_rect, .. } = self;
 
+        if state.image_id != image_id {
+            state.update(|state| state.image_id = image_id);
+        }
         if state.src_rect != src_rect {
             state.update(|state| state.src_rect = src_rect);
         }

